### ISSUE TITLE
[$500] BUG: Web - Selecting an Emoji on Emoji picker when navigated  via keyboard focuses on search input instead of selecting emoji reported by @aneequeahmad

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12043]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open web app(Compose box will be focused).
2. Press tab to navigate to emoji
3. Press enter to open Emoji picker and Search filed is focused .
4. Press tab to focus on emoji
5. Press enter to select emoji.

## Expected Result:
Emoji should be selected

## Actual Result:
Search box of emoji picker is focused

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.18-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/197010601-3657fd89-b50a-4f97-abc6-f699ba35c4b1.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @aneequeahmad
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666233818351049

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

https://user-images.githubusercontent.com/43996225/197010623-22f3aa6f-e1db-4053-898e-428ca29ae116.mp4

**Comment by mallenexpensify** [comment]
@maddylewis , new BZ chore SO is [here](https://stackoverflow.com/c/expensify/questions/14418), we want to ensure BZ stays assigned to all issues so we can get to BugZero,

**Comment by maddylewis** [comment]
reproducible here - 

https://user-images.githubusercontent.com/38016013/197410572-64cf2b53-591b-4e18-af99-eb237626bef0.mp4

**Comment by maddylewis** [comment]
job posted - https://www.upwork.com/jobs/~01e355bdb9800b63a9

**Comment by huzaifa-99** [comment]
### Proposal
This issue happens because `tab navigation` (with `Tab` key or `Shift+Tab` combo) is not handled for the `FlatList` in `EmojiPickerMenu/index.js` component. 
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L495-L510

So whenever `Tab` is pressed, this code block gets executed and sets the focus on the search input
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L167-L174

I believe there can be 2 approaches to solve this
1. We don't handle `tab navigation`, instead block it to overcome the current issue. This would require to update the `keyDownHandler` 
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L148-L175
by adding this code block before line `167`, that will just ignore `Tab` key press 
```js
if(keyBoardEvent.key === 'Tab') {
    keyBoardEvent.preventDefault();
    return;
}
```
Demo with blocked `Tab` key

https://user-images.githubusercontent.com/68777211/197425993-275d5ab7-5d91-45b5-93f8-07548533d6cb.mp4



2. We handle the `tab navigation` (`Tab` and `Shift+Tab`). Some of the logic to be used for this already exists in `highlightAdjacentEmoji` method
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L229-L321
 For this again we can can do it in 2 ways. 

2.1 We handle `Tab` navigation (both `Tab` and `Shift+Tab`) in a new method `highlightAdjacentEmojiOnTabNavigation` like below
```js
/**
 * Highlights emojis adjacent to the currently highlighted emoji depending on the tab key (with/without Shift Key)
 * @param {Boolean} isShiftKeyPressed
 */
highlightAdjacentEmojiOnTabNavigation(isShiftKeyPressed) {
    if (this.state.filteredEmojis.length === 0) {
        return;
    }

    // Tab without Shift enable navigation when search is focused
    if (this.searchInput && this.searchInput.isFocused() && !isShiftKeyPressed) {
        this.searchInput.blur(); 
    }

    // If nothing is highlighted and tab key is pressed (without Shift key)
    // select the first emoji
    if (this.state.highlightedIndex === -1 && !isShiftKeyPressed) {
        this.setState({highlightedIndex: this.firstNonHeaderIndex});
        this.scrollToHighlightedIndex();
        return;
    }

    let newIndex = this.state.highlightedIndex;
    const move = (steps, boundsCheck, onBoundReached = () => {}) => {
        if (boundsCheck()) {
            onBoundReached();
            return;
        }

        // Move in the prescribed direction until we reach an element that isn't a header
        const isHeader = e => e.header || e.spacer;
        do {
            newIndex += steps;
            if (newIndex < 0) {
                break;
            }
        } while (isHeader(this.state.filteredEmojis[newIndex]));
    };

    if(isShiftKeyPressed) {
        move(-1,
            () => this.state.highlightedIndex - 1 < this.firstNonHeaderIndex,
            () => {
                // Reaching start of the list, Shift + Tab set the focus to searchInput.
                this.focusInputWithTextSelect();
                newIndex = -1;
            }
        );
    } 
    else {
        move(1, () => this.state.highlightedIndex + 1 > this.state.filteredEmojis.length - 1);
    }

    // Actually highlight the new emoji and scroll to it if the index was changed
    if (newIndex !== this.state.highlightedIndex) {
        this.setState({highlightedIndex: newIndex});
        this.scrollToHighlightedIndex();
    }
}
```
and register this handler in the constructor (I would say after line `81`)
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L81
```js
this.highlightAdjacentEmojiOnTabNavigation = this.highlightAdjacentEmojiOnTabNavigation.bind(this);
```
and call it in the `keyDownHandler`
```js
if(keyBoardEvent.key === 'Tab') {
    keyBoardEvent.preventDefault();
    
    // Move the highlight when tab key is pressed (with/without Shift Key)
    this.highlightAdjacentEmojiOnTabNavigation(keyBoardEvent.shiftKey);
    return;
}
```

We would also have to completely block the `Shift` key if `Tab` key is not pressed. 
```js
if(keyBoardEvent.shiftKey || keyBoardEvent.key.startsWith('Shift')) {
    keyBoardEvent.preventDefault();
    return;
}
```
This is necessary because other wise a standalone `Shift` key press will cause issues with the tab navigation. On my local test, the standalone `Shift` key press caused the search input to focus, triggered by this conditional statement
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L167-L174  

We can also extract common code between the existing `highlightAdjacentEmoji` handler and the new `highlightAdjacentEmojiOnTabNavigation` handler into a separate method. (Please let me know if that's preferred)

Demo with tab navigation

https://user-images.githubusercontent.com/68777211/197426087-808cad07-a0c4-43d9-aa52-cdc28457e56f.mp4



2.2 The other approach involves the same flow as above but modifying the existing `highlightAdjacentEmoji` handler to work with `Tab` and `Shift+Tab` combo by adding some logic from the new `highlightAdjacentEmojiOnTabNavigation` handler. I am not sure if this would be preferred, because if we do then we would be handling key combo's in the same existing method `highlightAdjacentEmoji`.

Please let me know your thoughts. Thanks

**Comment by maddylewis** [comment]
I'll default to your expertise on this above proposal @marcaaron !

**Comment by marcaaron** [comment]
thanks, I'll chime in after @parasharrajat takes a look

**Comment by parasharrajat** [comment]
Thanks for the proposal @huzaifa-99.
> The other approach involves the same flow as above but modifying the existing highlightAdjacentEmoji handler to work with Tab and Shift+Tab combo by adding some logic from the new highlightAdjacentEmojiOnTabNavigation handler

IMO It is the best approach out of all you suggested.

But, is there a way of doing this via a focus/blur event on https://github.com/Expensify/App/blob/b9153fd780900f63b40b14e4ef3f5bf83e2d3184/src/components/EmojiPicker/EmojiPickerMenuItem.js#L27? 

I think it can be done the following way. 
1. Add focus/blur events on the `EmojiPickerMenuItem` and change the `highlightedIndex` as we do `onHover`. 
2. Pressing `Tab` will focus the next emoji and `shift+Tab` previous. 
3. Other things should work just fine. 

What do you think?

**Comment by huzaifa-99** [comment]
@parasharrajat Thanks for the feedback.

About the suggestion (**using focus/blur events**), it can be done. However I am not sure about these
1. Are we ok to proceed with modifying common components used between native and web apps? `EmojiPickerMenuItem` is also used in native `EmojiPickerMenu`
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.native.js#L11
2. Adding focus/blur events on `EmojiPickerMenuItem` will also add extra logic or computation in the same component that is only used for the web platform. Would this be an acceptable change? (more on this in the proposal details below)

If the above 2 points are ok to proceed with, please consider this proposal

### Proposal (Using Focus/Blur events)
Currently, there are 3 types of navigation in play here (to highlight/select an emoji)
a. Navigation using arrow keys
b. Navigation using mouse hover
c. Navigation using `Tab` or `Shift+Tab`

I would be referring these as Arrow-Nav, Mouse-Nav and Tab-Nav.

Let's say we implement Tab-Nav with Focus/Blur on `EmojiPickerMenuItem`. I propose these changes.
We would block `Tab` and `Shift` in `keyDownHandler`
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L148-L175
with something like this control flow (adding before like 167)
```js
if(
    keyBoardEvent.key === 'Tab' 
    || keyBoardEvent.shiftKey 
    || keyBoardEvent.key.startsWith('Shift')
) {
    return;
}
```
Blocking `Tab` and `Shift` would be necessary otherwise the logic below would run again and the focus would return to the search input when a `Tab` or `Shift` is pressed (I also mentioned this in previous proposal)
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L167-L174

Now the Tab-Nav works. Please see this demo
https://user-images.githubusercontent.com/68777211/198082871-13c8474b-d93a-4d54-bf55-f2709d401403.mp4

There are 2 issues now.
1. The focus adds a blue border (I am not sure if its by design or not. This border is also added to other elements that get focused by `Tab` like the search input above the emojis). Please let me know if this needs to be removed
2. The focused emoji is not highlighted. To add the highlight we will add onFocus and onBlur events in `EmojiPickerMenuItem` props
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenuItem.js#L9-L24
by adding these new props
```js
/** The function to call when an emoji is focused */
onFocus: PropTypes.func,

/** The function to call when an emoji is blured */
onBlur: PropTypes.func,
```
and in default props
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenuItem.js#L48-L52
by adding these default prop values
```js
onFocus: () => {},
onBlur: () => {},
```
and passing to `Pressable`
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenuItem.js#L27-L43
these prop values
```js
onFocus={props.onFocus}
onBlur={props.onBlur}
```
Next, we would have to listen to these callbacks in `EmojiPickerMenu/index.js`component
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L439-L452
by adding
```js
onFocus={() => this.setState({highlightedIndex: index})}
onBlur={() => this.setState({highlightedIndex: -1})}
```

Now the correct emoji is highlighted. Please see this updated demo
https://user-images.githubusercontent.com/68777211/198093067-4f511064-454f-4b28-adcf-fe7c4cfeebf3.mp4

#### Tab-Nav now works, but there are issues with other navigations (Mouse-Nav and Arrow-Nav) with the onFocus/onBlur change.
Please see this demo.
https://user-images.githubusercontent.com/68777211/198102696-28c5d408-7013-4e9c-8e6a-7beecde99aab.mp4

Basically `Tab` focus and `highlightedIndex` are out of sync.

#### These are the bugs introduced
**Bug-1)** We move to any emoji via `Tab`, **then hover mouse to a different emoji**. When `Tab` is pressed again the focus now again goes to the emoji immediately after the last emoji focused via `Tab`. (This is a bug that is introduced for Mouse-Nav)
**Bug-2)** We move to any emoji via `Tab`, **then use arrow keys to move to a different emoji**. When `Tab` is pressed again the focus now again goes to the emoji immediately after the last emoji focused via `Tab`. Also when the arrow keys are used again, they now start from the last focused emoji via `Tab` (This is a bug that is introduced for Arrow-Nav)

These bugs are happening because we are not blurring the emoji that was focused via `Tab` when we use other Navigation types (Mouse/Arrow). To overcome this issue we can add `ref` in `EmojiPickerMenuItem` to `blur` the emoji that just lost focus. Something like adding this
```js
const ref = useRef();

if(ref.current)
{
    if(props.isHighlighted) ref.current.focus();
    else ref.current.blur();
}
```
to `EmojiPickerMenuItem` and passing ref to `Pressable`
```js
 <Pressable
        ref={ref}
        ...
```

Updated demo with the `ref` changes

https://user-images.githubusercontent.com/68777211/198098936-c0948b9f-8ae8-4789-afed-009570f0e8cf.mp4

Now the `onFocus/onBlur` changes work as expected but I would not recommend this solution as It is computationally heavy than my previous proposal. 

I would tilt towards the previous proposal (because `ref` are not a good idea in my opinion, I may be wrong tho)

Please let me know what works best. Thanks

**Comment by parasharrajat** [comment]
Thanks for the detailed explanation. Focus/blur indeed looks too much overhead. I am ok with your previous approach. 

> 2.1 We handle Tab navigation (both Tab and Shift+Tab) in a new method highlightAdjacentEmojiOnTabNavigation like below

I looked at the existing code and you suggested an approach. It seems that keeping two functions will be more readable. We can of course dry the code as much as possible.

Questions: 
1. Will it not have the same issue as focus/blue with Arrow-Nav, Mouse-Nav?
2. I think we have to blur the focused element when Arrow-Nav, Mouse-Nav is used after Tab-nav to remove the blue border.

**Comment by huzaifa-99** [comment]
@parasharrajat. I agree about the dry part.

**Answers**
1. No. The reason being that we are just mimicking the `Tab` and `Shift+Tab`. When we press `Tab`, the same logic used for right arrow key press, gets executed (but in the new function with some other checkpoints) https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L298-L300 and when we press `Shift+Tab`, the logic used for left arrow key press get executed
https://github.com/Expensify/App/blob/70efb00eeb6e9617ecfa478e6348a252d4ac13f4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L289-L297
The handler is basically mapping Tab Press to Right Arrow Key Press and Shift+Tab Press to Left Arrow Key Press. The emoji selection works correctly as before with arrow keys
2. The blue border only show up when we do the real `Tab`, since we mimic the behaviour (as mentioned in point 1) the blue border does not appear (This is also illustrated in the demo of my original proposal https://github.com/Expensify/App/issues/12043#issuecomment-1288248326)

**Comment by parasharrajat** [comment]
Thanks for the answers. @huzaifa-99 's [proposal](https://github.com/Expensify/App/issues/12043#issuecomment-1288248326) option 

> 2.1 We handle Tab navigation (both Tab and Shift+Tab) in a new method highlightAdjacentEmojiOnTabNavigation like below 

looks good to me. 

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by marcaaron** [comment]
Can someone put a 2 sentence summary of the root cause and exactly what will be done to solve it? I am reading through these paragraphs of solutions and it all seems fairly complex for the feature want to support.

**Comment by marcaaron** [comment]
The whole `EmojiPickerMenu` event handling looks like a mess to me and is a bit hard to understand.

Why does this code not run when we press `Enter` ? 😕 

https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L160-L165

**Comment by huzaifa-99** [comment]
@marcaaron 

### Root Cause 
The `keydownhandler` 
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L148
listens to every key but only handles these
- All arrow keys (left, right, top, bottom)
- Enter key

If any other key (`Tab` in our case) is pressed then it just returns the focus to emoji search input
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L170

### Solution
I proposed (and @parasharrajat reviewed) to handle `Tab` and `Shift+Tab` keys as well (since they just return the focus back to the emoji search input)

**Comment by marcaaron** [comment]
>If any other key (Tab in our case) is pressed then it just returns the focus to emoji search input

But **we are pressing `'Enter'`** not Tab in the reproduction. So why is the event not handled?

>Press tab to focus on emoji

Works fine

>Press enter to select emoji

Moves focus to the search input. Why does this happen if the Enter event is handled?

Also, returning focus on "any" key press is kind of generic. Does anyone know why we return focus to the search input when **any other key is pressed**? I feel like we should not be returning the focus unless you tab all the way back to the search input or maybe some other accessibility reason I'm not familiar with.

**Comment by huzaifa-99** [comment]
@marcaaron because the `highlightedIndex` state is not updated when `Tab` is pressed. 

I understand your point that `Enter` is already handled. And since you pressed `Tab` and then `Enter`, this block (that selects the emoji on `Enter`) is not executed, because the `highlightedIndex` is `-1` (by default)
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L159-L165

For reference, the `highlightedIndex` is only updated when we use arrow keys or mouse hover. The `highlightAdjacentEmoji` on line 155 has the logic to update it via arrow keys
https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L149-L157

Apologies if my previous explanation was not to the point

**Comment by marcaaron** [comment]
Yeah so, I'm still kinda confused about why we would need to independently track the index of the emojis? We are listening to `Enter` presses on `document`, but my expectation would be that pressing `Enter` while an emoji is tabbed over by the browser would simulate clicking on it with the mouse. That's the typical browser behavior I'd think...

**Comment by marcaaron** [comment]
Why is this event here not firing?

https://github.com/Expensify/App/blob/e263ffa9f69e696790547bca4bf93d3864da90c4/src/components/EmojiPicker/EmojiPickerMenu/index.js#L441

**Comment by parasharrajat** [comment]
Global keydown listener is running in the Capture phase thus it fires first. And because tabbing does not change the internal index state. It does not work as expected. I suggested using `focus/blur` events earlier as I thought the default tabbing behavior will be enough to handle it. But due to customs handling of arrow keys and internal index state, it does not work well. It requires a few more changes as suggested in  https://github.com/Expensify/App/issues/12043#issuecomment-1292414925.

So I preferred manual management of Tab navigation same as other arrow and mouse-based to fit into the existing implementation.

**Comment by marcaaron** [comment]
I think if it's a problem to do both arrow and tab navigation then we should just pick one and stick with it.

Looks like Slack just makes their emoji picker not tab-able at all. That seems like a much simpler solution.

**Tab only activates these**

https://user-images.githubusercontent.com/32969087/198737171-5a8a1a04-76f3-48e3-ba23-6be0b3855cdd.mp4

**Arrows are only able to select emoji when input is focused**

https://user-images.githubusercontent.com/32969087/198737220-20fb88ba-b550-4972-9bb1-62c1535dbb6e.mp4

Feels like we should just copy this behavior instead of trying to make an `Enter` press work after a `Tab`. So, I think we should close this issue and create a new one that just disables tabbing to the emoji.

**Comment by huzaifa-99** [comment]
@marcaaron I had a similar recommendation in my original proposal comment https://github.com/Expensify/App/issues/12043#issuecomment-1288248326

> 1. We don't handle tab navigation, instead block it to overcome the current issue. This would require to update the keyDownHandler

The solution is to basically add this logic on the `keyDownHandler`
```js
if(keyBoardEvent.key === 'Tab') {
    keyBoardEvent.preventDefault();
    return;
}
```

**Comment by marcaaron** [comment]
Not great since we'd be preventing people from tabbing entirely.

**Comment by parasharrajat** [comment]
Should we change the expected behavior of the issue to meet the correct requirement?

**Comment by marcaaron** [comment]
@maddylewis what do you think and should we get more feedback or just go with my suggestion to not allow tabbing to an emoji? If we want to pivot I think it's ok to re-use this issue because the "actual behavior" is unexpected even if there's confusion about the expected behavior.

**Comment by eVoloshchak** [comment]
I proposed this for another [issue](https://github.com/Expensify/App/issues/11438#issuecomment-1296330930), but what do you think about
`Tab` - next emoji
`Shift+Tab` - previous emoji

https://user-images.githubusercontent.com/9059945/198897281-7f535854-2b2b-46d7-97e8-ebfa132e1ed4.mov

**Comment by marcaaron** [comment]
I think just skipping over emojis for the tabbing behavior would be the simplest solution. Feels like we are greatly exaggerating the importance of being able to tab to the emojis 🤷‍♂️

**Comment by mallenexpensify** [comment]
> Feels like we are greatly exaggerating the importance of being able to tab to the emojis 🤷‍♂️

I'm a keyboard nav fanatic but don't use tab much in Slack nor NewDot (I do in Spotify, they changed how it worked and it sux now).  My preference would be - once tab is hit twice (once to go from compose box to 😄 , second to highlight the first emoji) then the tab key should do what @eVoloshchak suggests above
> Tab - next emoji
Shift+Tab - previous emoji

**Comment by maddylewis** [comment]
are we still waiting on a finalized proposal for this one?

**Comment by michaelhaxhiu** [comment]
>I think just skipping over emojis for the tabbing behavior would be the simplest solution

Agreed with @marcaaron, and let's not forget we can always come back to incorporate "tab behavior" later on. My 2 cents here is that the keyboard navigation and keyboard selection (e.g. using directional arrows + enter) are the most important aspect of solving this.

**Comment by parasharrajat** [comment]
Ok, thanks. @michaelhaxhiu Please update the expected result section with
```
Prevent tabbing into the emojis. Users should not be able to tab onto the emoji list from the search input.
```

**Comment by mallenexpensify** [comment]
> I think just skipping over emojis for the tabbing behavior would be the simplest solution. Feels like we are greatly exaggerating the importance of being able to tab to the emojis 🤷‍♂️

and
> Prevent tabbing into the emojis. Users should not be able to tab onto the emoji list from the search input.

I think I misread Marc's comment earlier. I strongly disagree with this because `tab` + `enter` is the main way I access emoji.  If you can't access that way you have to always use a mouse, trackpad or a shortcut to open an emoji keyboard.   I think we should always be taking into keyboard accessibility for issues, even if we're not going down the 🐰🕳 now to address all potential accessibility issues.  

I'm gonna take a poll in #expensify-open-source to see if others use `tab` + `enter`.

**Comment by kidroca** [comment]
Removing tab navigation (or a particular item like the emoji button) is a workaround that reduces Apps usability for "shortcut users"

**Comment by mallenexpensify** [comment]
Ahh shit, I think I actually read it _right_ the first time and wrong the second (sorry...)
The point I was trying to make is that I think, when the cursor is in the compose box, the user should be able to click tab to highlight the emoji icon then hit enter to show search (where the cursor is now in the emoji search box).  

As far as tab _through_ all the emoji in the search box... I don't really care.  I just think you should be able to access the emoji search box by clicking tab then return. 

(def need that second cup o' coffee soon)

**Comment by kidroca** [comment]
Ok so the bug part is that we somewhat allow "Tab" navigation in the emoji picker (which **can** be opened by tabbing) 

We're considering whether to 
- prevent "Tab" navigation inside the emoji picker and therefore eliminate the bug
- implement correct "Tab" navigation 

Looks like you already agreed to go with 1), but IMO a better UX would be to still allow some partial tabbing e.g.
- you can press "Tab" to jump focus to the emojis, then browse with the keyboard (or just press enter to send the current emoji) 
- you can press "Tab" again to move focus back to the "search" field 
- you can omit pressing "Tab" and start browsing emojis (same as what we have now and Slack)

This behavior allows users to continue to use the "Tab" key and see that they can only use it to switch focus between the list and the search box, but to browse the list they see that they need to use something else 

E.g.
I open the emoji picker through tab navigation
I press tab again to go to the emoji list 
I press tab again thinking that I can browse emojis
I see this takes me back to search
I press tab again to go to the emoji list 
I try to use arrow keys to navigate

We can achieve that by:
- assign correct internal state when tab is pressed inside the emoji picker
- remove the native blue outline when an element is focused through tab and instead use the darker shade

**Comment by maddylewis** [comment]
@parasharrajat - when you have a chance, will you provide an update on when we think the PR for this fix will be ready?

**Comment by maddylewis** [comment]
doubled price of job here - https://www.upwork.com/jobs/~01e355bdb9800b63a9

**Comment by marcaaron** [comment]
>I think I misread Marc's comment earlier. I strongly disagree with this because tab + enter is the main way I access emoji.

I think we are overestimating the importance of this issue. I don't think Slack's millions of users care much about not being able to "tab to an emoji" and neither should ours.

**Comment by parasharrajat** [comment]
Marc has a point.

**Comment by maddylewis** [comment]
is someone able to provide a rundown of where we are with this one? are we moving forward with an existing proposal? or, deprioritizing this one? 

lmk - thanks!

**Comment by marcaaron** [comment]
The current proposals are complex and it's not clear that they are worth doing right now. I proposed an alternative solution to solve the bug which is to just stop allowing tab to emoji. I would prefer to update the issue description to stop allowing tabbing to emojis since that is the simplest way forward and seems totally acceptable to me. Broken down into it's parts the issue is really...

**Bug:** We are able to tab to an emoji but can't select it with an Enter key
**Feature request:** Make it so we can tab to emojis and select it with an Enter key

Based on the "bug" vs "feature request" conversations happening lately I'd suggest we fix the bug and not implement any new features then move on.

**Comment by aneequeahmad** [comment]
### PROPOSAL

As per discussion [above](https://github.com/Expensify/App/issues/12043#:~:text=The%20current%20proposals%20are%20complex%20and%20it%27s%20not%20clear%20that%20they%20are%20worth%20doing%20right%20now.%20I%20proposed%20an%20alternative%20solution%20to%20solve%20the%20bug%20which%20is%20to%20just%20stop%20allowing%20tab%20to%20emoji) `stop tab on emoji` is the fix.

**Solution:**
We can stop the tab when search input of Emoji menu is focused and tab is pressed. Following code is simplest to do the trick.

**Code changes:** 

Code should be added at the start of `this.keyDownHandler` listener.here after `line 148`
https://github.com/Expensify/App/blob/a4aa092bb912224d05460f60597b4102c8b5de94/src/components/EmojiPicker/EmojiPickerMenu/index.js#L148-L149


**Additional code:**

**Note:** Check of Search field is focued is added as it makes more sense IMO that search input is focused and tab is pressed then don't do anything on tab press. This is a kind of restriction of tab press after search field is focused and also future proof if any additonal functionaity or anything like that comes in.

```
 if(this.state.isFocused && this.searchInput.isFocused() && keyBoardEvent.key === 'Tab') {
                keyBoardEvent.preventDefault();
                return;
}
 ```

**Comment by parasharrajat** [comment]
@aneequeahmad What happens after the user has used arrow keys to highlight an emoji and then pressed Tab? I think it will focus on the emoji via Tab.

**Comment by aneequeahmad** [comment]
@parasharrajat Ahhhh yes, you're right this will happen. So instead of adding more checks its better to now allow tab at all on emoji picker since focus state is managed. This would be enough.

if(keyBoardEvent.key === 'Tab') {
                keyBoardEvent.preventDefault();
                return;
}

**Comment by aneequeahmad** [comment]
Also arrow keys can be used to focus on Search input again from emoji picker menu. 

So we should prevent Tab to work at all on emojiPicker.

**Comment by maddylewis** [comment]
@marcaaron - could you confirm if this issue falls into the recent announcement about [freezing any GH reports that relate to keyboard navigation or accessibility](https://expensify.slack.com/archives/C01GTK53T8Q/p1668604822482759)?

**Comment by parasharrajat** [comment]
@aneequeahmad https://github.com/Expensify/App/issues/12043#issuecomment-1295498807 https://github.com/Expensify/App/issues/12043#issuecomment-1295501141

**Comment by marcaaron** [comment]
It relates to keyboard navigation so I guess we should freeze it. Technically has a "bug" component because we are able to do something and it doesn't work at all as expected. @michaelhaxhiu can you help us out here?

**Comment by michaelhaxhiu** [comment]
Hm I'm conflicted 🤔 . I think we risk having to re-do this work in the future (when we fully revamp keyboard/navigation related issues) and so my gut tells me freeze it here. If you disagree though and think this is an easy fix (that we won't have to re-do or refactor) then it's probably kosher to proceed. 

I just think we need to make sure we are picking the right battles.

**Comment by parasharrajat** [comment]
I think it should be fine to do it now because we are going to disable the Tab on the emoji picker which will be the same in the future as well. But let's do whatever is best for the app. 

Decision is yours @michaelhaxhiu

**Comment by marcaaron** [comment]
I'm closing this as it doesn't feel like a priority right now and we are at an impasse. If we want to, we can report a new issue to propose that the emojis should not be "tabbable" or start a general conversation about how it should work.
